File: Surrounded Regions.cpp
class Solution {
public:
    void solve(vector<vector<char>> &board) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        
        if(board.empty())
            return;
        const int R = board.size();
        const int C = board[0].size();
        
        recordBound(board);
        for(int i=0; i<R; i++)
        {
            for(int j=0; j<C; j++)
                board[i][j] = board[i][j]=='#'?'O':'X';
        
        }
        
       
        
    }
    
    void recordBound(vector<vector<char>> &board)
    {
     
        const int R = board.size();
        const int C = board[0].size();
        
        for(int i=0; i<C; i++)
        {
            if(board[0][i]=='O')
                recordBound(board, 0, i, R, C);
            if(board[R-1][i]=='O')
                recordBound(board, R-1, i, R, C);
        }
        
        for(int i=1; i<R-1; i++)
        {
            if(board[i][0]=='O')
                recordBound(board, i, 0, R, C);
            if(board[i][C-1]=='O')
                recordBound(board, i, C-1, R, C);
        }
    }
    
    void recordBound(vector<vector<char>> &board, int row, int colum, const int R, const int C)
    {
        if(row<0||row==R||colum<0||colum==C)
            return;
        else  if(board[row][colum]=='X'||board[row][colum]=='#')
            return;
        else
        {
            board[row][colum] = '#';
            recordBound(board, row, colum+1, R, C);
            recordBound(board, row, colum-1, R, C);
            recordBound(board, row-1, colum, R, C);
            recordBound(board, row+1, colum, R, C);
        }
       
    }
};

File: Surrounded Regions.java
public class Solution {
    public void solve(char[][] board) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if(board.length == 0 || board[0].length == 0) return;
        Queue<Integer> q = new LinkedList<Integer>();
        int iLen = board.length;
        int jLen = board[0].length;
        for(int j = 0; j < jLen; j++)
        {
            if(board[0][j]=='O')
            {
                q.offer(j);
                board[0][j]='+';
            }
            if(board[iLen-1][j]=='O')
            {
                q.offer((iLen-1)*jLen+j);
                board[iLen-1][j]='+';
            }
        }
        for(int i = 1; i < iLen-1; i++)
        {
            if(board[i][0]=='O')
            {
                q.offer(i*jLen);
                board[i][0]='+';
            }
            if(board[i][jLen-1]=='O')
            {
                q.offer(i*jLen+jLen-1);
                board[i][jLen-1]='+';
            }
        }
        
        
        while(!q.isEmpty())
        {
            int cur = q.poll();
            int t = cur / board[0].length;
            int s = cur % board[0].length;
            visit(t - 1, s, q, board);
            visit(t, s - 1, q, board);
            visit(t + 1, s, q, board);
            visit(t, s + 1, q, board);
        }
        
        for(int i = 0; i < iLen; i++)
        {
            for(int j = 0; j < jLen; j++)
            {
                if(board[i][j] == 'O') board[i][j] = 'X';
                else if(board[i][j] == '+') board[i][j] = 'O';
            }
        }
    }
    
    private void visit(int i, int j, Queue<Integer> q, char[][] board)
    {
        int m = board.length, n = board[0].length;
        if(i < 0 || i >= m || j < 0 || j >= n || board[i][j] != 'O') return;
        board[i][j] = '+';
        q.offer(i * n + j);
    }
}
